x/delayedack/keeper/rollapp_packet.go

```diff
@@ -0,0 +1,86 @@
+func (k Keeper) SetRollappPacket(ctx sdk.Context, rollappID string, rollappPacket types.RollappPacket) {
+	logger := ctx.Logger()
+	logger.Debug("Saving rollapp packet", "rollappID", rollappID, "channel", rollappPacket.Packet.DestinationChannel,
+		"sequence", rollappPacket.Packet.Sequence, "proofHeight", rollappPacket.ProofHeight)
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.RollappPacketKeyPrefix))
+	b := k.cdc.MustMarshal(&rollappPacket)
+	store.Set(types.GetRollappPacketKey(
+		rollappID,
+		types.RollappPacket_PENDING,
+		rollappPacket.ProofHeight,
+		*rollappPacket.Packet,
+	), b)
+}
+
+// UpdateRollappPacketStatus deletes the current rollapp packet and creates a new one with and updated status under a new key.
+// It assumes that the packet has been previously stored with the pending status.
+func (k Keeper) UpdateRollappPacketStatus(ctx sdk.Context, rollappID string, rollappPacket types.RollappPacket, newStatus types.RollappPacket_Status) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.RollappPacketKeyPrefix))
+
+	// Delete the old rollapp packet
+	oldKey := types.GetRollappPacketKey(rollappID, types.RollappPacket_PENDING, rollappPacket.ProofHeight, *rollappPacket.Packet)
+	store.Delete(oldKey)
+
+	// Update the packet
+	rollappPacket.Status = newStatus
+
+	// Create a new rollapp packet with the updated status
+	newKey := types.GetRollappPacketKey(rollappID, newStatus, rollappPacket.ProofHeight, *rollappPacket.Packet)
+	b := k.cdc.MustMarshal(&rollappPacket)
+	store.Set(newKey, b)
+}
+
+// ListRollappPendingPackets retrieves a list of pending rollapp packets from the KVStore.
+// It builds a prefix using the rollappID and the pending status, and iterates over the range from lastProofHeight to proofHeight.
+// If the packet's proofHeight is less than or equal to the maxProofHeight, it is added to the list.
+// The function returns the list of pending packets.
+func (k Keeper) ListRollappPendingPackets(
+	ctx sdk.Context,
+	rollappId string,
+	maxProofHeight uint64,
+) (list []types.RollappPacket) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.RollappPacketKeyPrefix))
+
+	// Build the prefix which is composed of the rollappID and the status
+	var prefix []byte
+	prefix = append(prefix, []byte(rollappId)...)
+	prefix = append(prefix, []byte("/")...)
+
+	// Append the pending status to the prefix
+	pendingStatusBytes := []byte(fmt.Sprint(types.RollappPacket_PENDING))
+	prefix = append(prefix, pendingStatusBytes...)
+	prefix = append(prefix, []byte("/")...)
+
+	// Iterate over the range from lastProofHeight to proofHeight
+	iterator := sdk.KVStorePrefixIterator(store, prefix)
+	defer iterator.Close() // nolint: errcheck
+
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.RollappPacket
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		keyParts := strings.Split(string(iterator.Key()), "/")
```

Comment:
why u need to extract height it from the key? u have it in the value

Reply:
true!